### ISSUE TITLE
GELF codec support

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#4868]
Support GELF as a codec for use when sending data to Graylog: https://docs.graylog.org/en/3.3/pages/sending/gelf.html

GELF can be sent over HTTP, UDP, and TCP to Graylog.

We should add `gelf` as a codec option for both sources and sinks.

Example configuration:

```toml
[sources.foo]
type = "stdin"
decoding.codec = "gelf"
```

```toml
[sinks.bar]
type = "console"
encoding.codec = "gelf"
```

**Comment by karlmartink** [comment]
This would be very beneficial.

**Comment by dm3ch** [comment]
Hi, Guys!
I'm researching Vector usage for one of the companies I work for. Currently we have a lot of components that are using GELF as a format of messaging, so for us it was a blocker.
As said on official graylog page https://docs.graylog.org/en/3.1/pages/gelf.html gelf is a format build based on json and have different possible transports (TCP/UDP/HTTP).
Keeping this in mind I have decided to test if vector currently able to send/recieve GELF messages, and here's results of my research.

For tests I have been using:
Vector: 0.11.1
Graylog: v4.0.2+1987d10

Sending messages from vector to GELF reciever:
My test config:
```
[sources.stdin]
  type = "stdin"

[transforms.gelf]
  # General
  type = "add_fields" # required
  inputs = ["stdin"] # required
  overwrite = true # optional, default

  # Fields
  fields.version = "1.1"
  fields.short_message = "Short message: {{ message }}"
  fields.full_message = "Full message: {{ message }}"
  fields.level = 3
  fields._foo = "bar"


[sinks.http]
  # General
  type = "http" # required
  inputs = ["gelf"] # required
  compression = "gzip" # optional, default
  uri = "http://graylog-tcp:12201/gelf" # required

  # Batch
  batch.max_bytes = 1049000 # optional, default, bytes
  batch.timeout_secs = 1 # optional, default, seconds

  # Encoding
  encoding.codec = "ndjson" # required
  encoding.except_fields = ["source_type", "message"]
  encoding.timestamp_format = "unix"

  # Healthcheck
  # healthcheck.enabled = true # optional, default

[sinks.tcp]
  # General
  type = "socket" # required
  inputs = ["gelf"] # required
  address = "graylog-tcp:12202" # required, required when mode = `tcp` or `udp`
  mode = "tcp" # required

  # Encoding
  encoding.codec = "json" # required
  encoding.except_fields = ["source_type", "message"]
  encoding.timestamp_format = "unix"

  # Healthcheck
  #healthcheck.enabled = true # optional, default

[sinks.udp]
  # General
  type = "socket" # required
  inputs = ["gelf"] # required
  address = "graylog-udp:12203" # required, required when mode = `tcp` or `udp`
  mode = "udp" # required

  # Encoding
  encoding.codec = "json" # required
  encoding.except_fields = ["source_type", "message"]
  encoding.timestamp_format = "unix"

  # Healthcheck
  #healthcheck.enabled = true # optional, default

```
- HTTP - *Works* I have setup GELF HTTP input with default configuration. I was able to send messages to the graylog and I saw them in graylog. `ndjson` encoding worked for me, `json` - not. Also I found that GZIP compression also works.
- TCP - *Not works* I have setup GELF TCP input with default configuration. I was able to send messages to the graylog, but suddenly I got only one message (and only after vector shutdowm), this is caused cause graylog expects special delimeter byte after each message (\0 or \n), unfortunately right now I haven't found way to configure this in vector socket sync.
- UDP - *Works* I have setup GELF UDP input with default configuration. I was able to send messages to the graylog and I saw them in graylog.

Recieving GELF messages by vector:
My test config:
```
[sources.gelf_http]
  type = "http" # required
  address = "0.0.0.0:8001" # required
  encoding = "ndjson" # optional, default

[sources.gelf_tcp]
  type = "socket" # required
  address = "0.0.0.0:8002" # required, required when mode = `tcp` or `udp`
  max_length = 102400 # optional, default, bytes
  mode = "tcp" # required

[sources.gelf_udp]
  type = "socket" # required
  address = "0.0.0.0:8003" # required, required when mode = `tcp` or `udp`
  max_length = 102400 # optional, default, bytes
  mode = "udp" # required

[sinks.console]
  # General
  type = "console" # required
  inputs = ["gelf_http", "gelf_tcp", "gelf_udp"] # required
  target = "stdout" # optional, default

  # Encoding
  encoding.codec = "json" # required
```
I tried to add some transform to each type of source to understand if vector correctly parses message.
- HTTP - *Works*
- TCP - *Works*
- UDP - *Works*

Conclusion:
* Currently vector is possible to recieve/transform/send GELF messages, but with some limitations.
* GELF iteroperability requires further testing, cause I haven't yet tested described above configurations on real cases (Lot's of long messages).
* For my opinion, the main current pain with GELF is additional fields - in GELF all additional fields should be prefixed with _, so to transform message with additional fields to GELF remap of all additional fields is required. It's possible to do that with remap, but you need to specify it for each field. I assume that using lua transform for that is possible, but I haven't yet tested that.
* It would be better to introduce separate GELF sink/source to make configuration easier.
* Maybe it would be a good idea to add delimeter byte option (disabled by default) to the socket sink, this would allow us to use TCP transport for GELF messages.
* I haven't tested Kafka transport yet, but based on other test results I assume that it would work.

**Comment by dm3ch** [comment]
@binarylogic Sorry for mention. I would like to try to adding basic GELF codec support (for no without compression and other advanced features, just something like ndjson but with fields renaming).

Suddenly I spent about an hour and didn't found where does actually codec code lives. Also I still haven't understood why/how some thinks limiting codecs usage (http could use ndjson, but socket could use only plain json).
Is there any documentation that could be helpful for newbies (from dev perspective)?

**Comment by binarylogic** [comment]
Hi @dm3ch, unfortunately, we don't have any public documentation on codecs. The reason HTTP uses "ndjson" is that the final payload is encoded as such (newline delimited JSON). The socket source uses plain old "json" since it is decoding one event at a time. Hope that helps clarify the naming. Feel free to hop in our [chat](https://chat.vector.dev) if you need further development help.

**Comment by eplightning** [comment]
Graylog's native HTTP input doesn't seem work very well with vector's `http` sink. Graylog doesn't support any message batching and in general it looks like the support for HTTP input is afterthought (https://github.com/Graylog2/graylog2-server/pull/5924), not something supported as well as TCP or UDP.

If anyone using Graylog wants to give vector a try I have written simple log forwarder that accepts input from vector (via http or vector sink) and forwards them to Graylog via TCP or UDP.
https://github.com/eplightning/gelf-forwarder

That's obviously not an ideal solution but I think it's an acceptable workaround until vector gets native GELF support.

**Comment by jszwedko** [comment]
Just noting in case anyone picks this up, I think we can model this as another "codec" similar to the `syslog` codec. That would mean adding it as an encoder and decoder here: https://github.com/vectordotdev/vector/tree/master/lib/codecs

**Comment by mpfz0r** [comment]
> Graylog doesn't support any message batching

This has been addressed since Graylog 4.3
https://github.com/Graylog2/graylog2-server/pull/11529

**Comment by neuronull** [comment]
> > Graylog doesn't support any message batching
> 
> This has been addressed since Graylog 4.3 [Graylog2/graylog2-server#11529](https://github.com/Graylog2/graylog2-server/pull/11529)

#13292 was opened to address batching / chunking.

**Comment by JokerQyou** [comment]
I thought basic GELF support was already merged, but it seems 0.23.3 does not support `decoding.codec = "gelf"` in sinks?

**Comment by neuronull** [comment]
> I thought basic GELF support was already merged, but it seems 0.23.3 does not support `decoding.codec = "gelf"` in sinks?

The sink side is "encoding" :

```toml
[sinks.sink0.encoding]
codec = "gelf"

```

**Comment by JokerQyou** [comment]
> The sink side is "encoding" :
> 
> ```toml
> [sinks.sink0.encoding]
> codec = "gelf"
> ```

Sorry, what I meant was in sources. My basic usage is like this: my existing apps write logs using GELF via UDP, so I want vector to be able to listen on a specific address and collect those logs, then send them to a central logging server using JSON. With 0.23.3 is it possible?

**Comment by jszwedko** [comment]
Hi @JokerQyou ! This change will be included in the forthcoming 0.24.0 release. You can try out the nightly builds if you want to play with it ahead of the release.

**Comment by arch-xtof** [comment]
Hello! I am trying to use GELF input on Vector `0.24.1`. My main goal is to forward logs from Graylog to Vector. I have created a stream along with output in Graylog and my Vector config is the following:
```
  [sources.gelf_stream]
    type = "socket"
    address = "0.0.0.0:9091"
    mode = "tcp"
    decoding.codec = "gelf"

  [sinks.term]
    type = "console"
    inputs = [ "gelf_stream" ]
    target = "stdout"

    [sinks.term.encoding]
      codec = "text"
```

Vector does not seem to recognize the **"\0"** delimiter of GELF and considers all accumulated content thought-out the TCP session as one message. Due to this, Vector fails to decode the message. The same would happen if I follow the [Graylog docs on sending GELF via netcat](https://archivedocs.graylog.org/en/3.1/pages/gelf.html#sending-gelf-messages-via-tcp-using-netcat) and I would get the following warning:
`Sep 28 11:00:51 xxx vector[27598]: 2022-09-28T11:00:51.187830Z  WARN source{component_kind="source" component_id=gelf_stream component_type=socket component_name=gelf_stream}:connection{peer_addr=127.0.0.1:38964}: vector::internal_events::decoder: Failed deserializing frame. error=trailing characters at line 1 column 113 internal_log_rate_secs=10`
